### PR TITLE
Improve testing, handle isNull, and improve between pruning.

### LINKED ISSUE #1324: PruningFilters crash on IS NULL queries
E.g. the RangeFilter in this first cast already

```
  bool can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
                 const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const override {
    const auto value = type_cast_variant<T>(variant_value);
```

they should probably just return false for predicates that they know nothing about.

### PR DESCRIPTION
**Pull request by Bouncner**
Fixes #1324.

Thing done:
 * Add tests for unsupported predicates. They have been skipped before and returned false, which is fine, but NULL_VALUE parameters have not been handled (which can occur when the predicate is isNull). In these cases, the method crashed as it could not convert the value.
 * To fix the error, NULL parameters are now checked and return false directly.
 * Add tests for between and improve between pruning, which only pruned a few cases before.

### CODE REVIEW

File: src/test/statistics/chunk_statistics/range_filter_test.cpp [hunk 1]
```diff
@@ -207,6 +222,26 @@ TYPED_TEST(RangeFilterTest, CanPruneOnBounds) {
...
+  EXPECT_TRUE(filter->can_prune(PredicateCondition::Between, {-3000}, {-2000}));
+  EXPECT_TRUE(filter->can_prune(PredicateCondition::Between, {-999}, {1}));
+  EXPECT_TRUE(filter->can_prune(PredicateCondition::Between, {104}, {1004}));
+  EXPECT_TRUE(filter->can_prune(PredicateCondition::Between, {10'000'000}, {20'000'000}));
+
+  EXPECT_FALSE(filter->can_prune(PredicateCondition::Between, {-3000}, {-500}));
+  EXPECT_FALSE(filter->can_prune(PredicateCondition::Between, {101}, {103}));
+  EXPECT_FALSE(filter->can_prune(PredicateCondition::Between, {102}, {1004}));
+
+  // SQL between in inclusive
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: SQL between IS inclusive. I wasn't sure myself what the default behaviour is, so I thought I save the reader some time.

Fixed.

File: src/test/statistics/chunk_statistics/range_filter_test.cpp [hunk 2]
```diff
@@ -167,6 +167,21 @@ TYPED_TEST(RangeFilterTest, MoreRangesThanValues) {
   EXPECT_TRUE(filter->can_prune(PredicateCondition::GreaterThan, {this->_max_value}));
 }
 
+// create more ranges than distinct values in the test data
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: What do you mean?

File: src/test/statistics/chunk_statistics/range_filter_test.cpp [hunk 3]
```diff
@@ -207,6 +222,26 @@ TYPED_TEST(RangeFilterTest, CanPruneOnBounds) {
   EXPECT_TRUE(filter->can_prune(PredicateCondition::GreaterThan, {this->_after_range}));
 }
 
+// Test larger value ranges.
+TYPED_TEST(RangeFilterTest, Between) {
+  const auto filter = RangeFilter<TypeParam>::build_filter(this->_values);
+
+  // This one has bounds in gaps, but cannot prune.
+  EXPECT_FALSE(filter->can_prune(PredicateCondition::Between, {this->_max_value - 1}, {this->_after_range}));
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: As far as I know, yes. Happy to take suggestions how to avoid the `this`.

File: src/lib/statistics/chunk_statistics/range_filter.hpp [hunk 4]
```diff
@@ -66,9 +71,43 @@ class RangeFilter : public AbstractFilter {
...
+
+        // Larger than the segment's maximum.
+        if (can_prune(PredicateCondition::GreaterThanEquals, std::min(value, value2))) {
+          return true;
+        }
+
+        // Get value range or next larger value range if searched value is in a gap.
+        auto start_lower = std::lower_bound(
+            std::begin(_ranges), std::end(_ranges), value,
+            [](std::pair<T, T> range, T compare_value) -> bool { return range.second < compare_value; });
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/statistics/chunk_statistics/range_filter.hpp]: Whoops. Yes, thanks.

File: src/test/statistics/chunk_statistics/range_filter_test.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrks) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: tbh, I haven't really understood when you would need it and when not.

File: src/test/statistics/chunk_statistics/range_filter_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrks) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: Your test is supposed to do something to unsupported predicates. Why does it matter how many ranges you have?

[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: You marked the `MoreRangesThanValues` test, which does not check for unsupported predicates. `DoNotPruneUnsupportedPredicates` does check for unsupported and only builds a single default-sized filter.

Is that ok?

File: src/test/statistics/chunk_statistics/range_filter_test.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: As far as I understood it's caused by the way Gtest expands `TYPED_TEST`s to create typed test cases. For normal test cases, `this` ain't necessary.

File: src/test/statistics/chunk_statistics/range_filter_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrks) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: The comment is on top of `RangeFilterTest.DoNotPruneUnsupportedPredicates`

File: src/lib/statistics/chunk_statistics/min_max_filter.hpp [hunk 5]
```diff
@@ -18,7 +18,13 @@ class MinMaxFilter : public AbstractFilter {
 
   bool can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
                  const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const override {
+    // Early exit for NULL variants.
+    if (variant_is_null(variant_value)) {
```
Comment (mrzzzrm) [inline src/lib/statistics/chunk_statistics/min_max_filter.hpp]: Wouldn't the safer (additional) condition be `if predicate_type is IsNull or IsNotNull`? Because that's the operation the the Filter(s) have no idea about, right? variant_value being NULL for these conditions seems to be more of an indirect side-effect.

File: src/lib/statistics/chunk_statistics/range_filter.hpp [hunk 6]
```diff
@@ -66,9 +71,43 @@ class RangeFilter : public AbstractFilter {
...
+        // Larger than the segment's maximum.
+        if (can_prune(PredicateCondition::GreaterThanEquals, std::min(value, value2))) {
+          return true;
+        }
+
+        const auto range_comp = [](std::pair<T, T> range, T compare_value) -> bool {
+          return range.second < compare_value;
+        };
+        // Get value range or next larger value range if searched value is in a gap.
+        const auto start_lower = std::lower_bound(std::begin(_ranges), std::end(_ranges), value, range_comp);
```
Comment (mrzzzrm) [inline src/lib/statistics/chunk_statistics/range_filter.hpp]: _ranges.begin()/end() is shorter

File: src/test/statistics/chunk_statistics/range_filter_test.cpp (hunk 3 above)
Comment (mrzzzrm) [inline src/test/statistics/chunk_statistics/range_filter_test.cpp]: You tested that `BETWEEN this->_max_value - 1 AND this->_after_range` can not be pruned - good! Why not do due diligence and test  `BETWEEN this->_max_value AND this->_after_range` and  `BETWEEN this->_max_value + 1 AND this->_after_range` as well? Why not do the same for the lower end of the value range? And for one or two gaps?

Comment (mrzzzrm) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: Having those tests would scream "I am properly tested for edge cases!!!" much more than a billion randomly generated values :P :P

**Comment by mrks** [comment]
I'll be happy once @mrzzzrm is.

File: src/lib/statistics/chunk_statistics/min_max_filter.hpp (hunk 5 above)
Comment (Bouncner) [reply src/lib/statistics/chunk_statistics/min_max_filter.hpp]: The current approach also ensures that a predicate such as `a = NULL` does not crash.
I could check for the predicate type, but this would be three lines of code instead of the default value in the switch.

But you reminded me that we might want to check for value2 to be non-null as well.

I can live with any solution.

Comment (Bouncner) [reply src/lib/statistics/chunk_statistics/min_max_filter.hpp]: But you reminded me that we might want to check for value2 to be non-null as well. But I don't like that to be in any statistics component either.

Just ignore such malformed cases or should that be part of the optimizer / parser?

File: src/test/statistics/chunk_statistics/range_filter_test.cpp (hunk 3 above)
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: I already have better tests as part of #1321, but I don't know when I will finish the PR. I'll copy your comment to the issue.

File: src/lib/statistics/chunk_statistics/min_max_filter.hpp (hunk 5 above)
Comment (mrzzzrm) [reply src/lib/statistics/chunk_statistics/min_max_filter.hpp]: I'd just do `if (is_variant_null(value) or is_variant_null(value2) or predicate_type is IS (NOT) NULL) return false-Filter-dont-know-how-to-handle-this`, I think that's the least-code approach to ensuring that nothing goes wrong.

Comment (Bouncner) [reply src/lib/statistics/chunk_statistics/min_max_filter.hpp]: Do you think we should return false for `can_prune(PredicateCondition::LessThan, {5}, NULL_VALUE)` (assume 5 can be pruned)?

Comment (mrzzzrm) [reply src/lib/statistics/chunk_statistics/min_max_filter.hpp]: (Don't worry about it too much, the new CardinalityEstimator will likely wrap PredicateCondition+Value+Value2 into a single data structure and ensure that illegal expressions aren't possible)

Apart from that: yes, I'd say so. If the PredicateCondition is so that value2 can be ignored, then it is valid for the Filter to ignore it, whatever its value is.

Comment (mrzzzrm) [reply src/lib/statistics/chunk_statistics/min_max_filter.hpp]: But if you use the check as I proposed (`if (is_variant_null(value) or is_variant_null(value2) or predicate_type is IS (NOT) NULL)`) and return false, then that would be okay with me as well. It's the callers responsibility to formulate a valid expression.

Comment (Bouncner) [reply src/lib/statistics/chunk_statistics/min_max_filter.hpp]: I changed the code accordingly. Ready to merge?
